### PR TITLE
Add zrest authentication support

### PR DESCRIPTION
**Pull request by sysvinit**
These commits add and document a `--zrest` option which enables the "zrest" authentication scheme supported by [restund](https://github.com/wireapp/restund), which derives passwords based on the HMAC of the username using a secret shared between the TURN server and an external application. 

This functionality is very similar to the existing "TURN REST API" support in coturn (which is an implementation of [I-D.uberti-behave-turn-rest](https://datatracker.ietf.org/doc/html/draft-uberti-behave-turn-rest-00)), so the additions in this PR are mostly based on the existing code. In particular, the zrest authentication code reuses the existing authentication secrets which can be configured either using `--static-auth-secret` or in an external database.

When the `--zrest` flag is enabled, it will turn on the TURN REST API code, but use the zrest-specific username and password handling code paths.

### CODE REVIEW

File: src/apps/relay/userdb.c [hunk 1]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+		uint32_t kidx;
+		turn_time_t deadline, now;
+		secrets_list_t sl;
+		unsigned char hmac[MAXSHASIZE];
+		unsigned int hmac_len;
+		char *pwd;
+		size_t sll = 0, pwd_len;
+		password_t pwdtmp;
+
+		if(zrest_validate_username((char *)usname, &kidx, &deadline)!=0)
```
Comment (supersven) [inline src/apps/relay/userdb.c:624]:
```suggestion
		if(zrest_validate_username((char *)usname, &kidx, &deadline) != 0)
```
Formatting.

File: src/apps/relay/userdb.c [hunk 2]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+
+		now = (turn_time_t) time(NULL);
+		if (!turn_time_before(now, deadline)) {
+			TURN_LOG_FUNC(TURN_LOG_LEVEL_WARNING, "zrest authentication token expired\n");
+			return ret;
+		}
+
+		init_secrets_list(&sl);
+
+		if(get_auth_secrets(&sl, realm)<0)
```
Comment (supersven) [inline src/apps/relay/userdb.c:635]:
```suggestion
		if(get_auth_secrets(&sl, realm) < 0)
```
Formatting.

File: src/apps/relay/userdb.c [hunk 3]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+			return ret;
+
+		for(sll=0;sll<get_secrets_list_size(&sl);++sll) {
```
Comment (supersven) [inline src/apps/relay/userdb.c:638]:
```suggestion
		for(sll=0; sll < get_secrets_list_size(&sl); ++sll) {
```
Formatting.

File: src/apps/relay/userdb.c [hunk 4]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+			const char* secret = get_secrets_list_elem(&sl, sll);
+
+			if(!secret)
+				continue;
+
+			if(stun_calculate_hmac(usname, strlen((char*)usname), (const uint8_t *)secret, strlen(secret), hmac, &hmac_len, SHATYPE_SHA512)!=0)
```
Comment (supersven) [inline src/apps/relay/userdb.c:644]:
```suggestion
			if(stun_calculate_hmac(usname, strlen((char*)usname), (const uint8_t*)secret, strlen(secret), hmac, &hmac_len, SHATYPE_SHA512) != 0)
```
Formatting.

File: src/apps/relay/userdb.c [hunk 5]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+
+			pwd = base64_encode(hmac, hmac_len, &pwd_len);
+			if(!pwd)
+				continue;
+			if(pwd_len < 1) {
+				free(pwd);
+				continue;
+			}
+
+			if(stun_produce_integrity_key_str((uint8_t *)usname, realm, (uint8_t *)pwd, key, SHATYPE_DEFAULT)!=0) {
```
Comment (supersven) [inline src/apps/relay/userdb.c:655]:
```suggestion
			if(stun_produce_integrity_key_str((uint8_t*) usname, realm, (uint8_t*) pwd, key, SHATYPE_DEFAULT) != 0) {
```
Formatting.

File: src/apps/relay/userdb.c [hunk 6]
```diff
@@ -532,6 +611,66 @@ int get_user_key(int in_oauth, int *out_oauth, int *max_session_time, uint8_t *u
...
+				free(pwd);
+				continue;
+			}
+
+			if(stun_check_message_integrity_by_key_str(TURN_CREDENTIALS_LONG_TERM, ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh), key, pwdtmp, SHATYPE_DEFAULT)<1) {
```
Comment (supersven) [inline src/apps/relay/userdb.c:660]:
```suggestion
			if(stun_check_message_integrity_by_key_str(TURN_CREDENTIALS_LONG_TERM, ioa_network_buffer_data(nbh), ioa_network_buffer_get_size(nbh), key, pwdtmp, SHATYPE_DEFAULT) < 1) {
```
Formatting.

File: src/apps/relay/userdb.c [hunk 7]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+{
+	regex_t reg;
+	regmatch_t matches[ZREST_USERNAME_CAPTURES];
+	int ret;
+	char *span, *tmp, *endptr;
+	regoff_t span_len;
+
+	ret = -1;
+
+	/* XXX(sysvinit): this regex should not be repeatedly compiled, this
```
Comment (supersven) [inline src/apps/relay/userdb.c:418]: If it's so inefficient, why not solve this todo and compile the regex pattern in a static initialization block? :thinking:

**Review by supersven — APPROVED** [review]
Despite some minor things, this looks really great!

File: src/apps/relay/userdb.c [hunk 8]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+	   could be very inefficient under load. */
+	if(regcomp(&reg, ZREST_USERNAME_REGEX, REG_EXTENDED)!=0) {
+		TURN_LOG_FUNC(TURN_LOG_LEVEL_ERROR, "Could not compile zrest authentication regex\n");
+		return ret;
+	}
+
+	if(regexec(&reg, usname, ZREST_USERNAME_CAPTURES, matches, 0)!=0) {
```
[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c:425]: See above.

File: src/apps/relay/userdb.c [hunk 9]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+		TURN_LOG_FUNC(TURN_LOG_LEVEL_INFO, "Username does not match token format: %s\n", usname);
+		goto regout;
+	}
+
+	if(matches[ZREST_DEADLINE_CAP].rm_so == -1 || matches[ZREST_KEYINDEX_CAP].rm_so == -1) {
+		TURN_LOG_FUNC(TURN_LOG_LEVEL_ERROR, "Empty capture groups in valid zrest username\n");
+		goto regout;
+	}
+
+	/* XXX this assumes that sizeof(int) == 4; turn_time_t is an alias for int32_t */
```
[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c:435]: The "key index" in the authentication token is a 32-bit integer. If one assumes that `int` is 4 bytes wide, then this means that we can use `strtod()` to read a 32-bit integer. I don't know how well this assumption holds up in practice -- while I think POSIX says that `char` is always 8 bits wide, I don't know whether it says anything about integer types.

That being said, this also misses the fact that the key index is an unsigned integer -- on second thoughts, there might be a better way to go about doing this. I'll take another look at this.

File: src/apps/relay/userdb.c [hunk 10]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+	span_len = matches[ZREST_KEYINDEX_CAP].rm_eo - matches[ZREST_KEYINDEX_CAP].rm_so;
+	tmp = strndup(span, (size_t) span_len);
+	if(!tmp)
+		goto regout;
+
+	errno = 0;
+	*keyindex = (uint32_t) strtod(span, &endptr);
+	if(!endptr)
+		goto tmpout;
+	if(errno!=0)
```
[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c:447]: See above.

File: src/apps/relay/userdb.c [hunk 11]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+	span_len = matches[ZREST_DEADLINE_CAP].rm_eo - matches[ZREST_DEADLINE_CAP].rm_so;
+	tmp = strndup(span, (size_t) span_len);
+	if(!tmp)
+		goto regout;
+
+	errno = 0;
+	*deadline = (turn_time_t) strtod(span, &endptr);
+	if(!endptr)
+		goto tmpout;
+	if(errno!=0)
```
[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c:464]: See above.

File: src/apps/relay/userdb.c (hunk 1 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:624]: See above.

File: src/apps/relay/userdb.c (hunk 2 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:635]: See above.

File: src/apps/relay/userdb.c (hunk 3 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:638]: See above. I don't like how dense this is, but it's what the existing code does, and I'd rather not diverge within the same source file.

File: src/apps/relay/userdb.c (hunk 4 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:644]: See above.

File: src/apps/relay/userdb.c (hunk 5 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:655]: See above.

File: src/apps/relay/userdb.c (hunk 6 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:660]: See above.

File: src/apps/relay/userdb.c (hunk 7 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c:418]: I was mostly running with "first make it work, then make it work fast", though you're right about initialising this statically if possible. I'll take another look at this later.

File: src/apps/relay/userdb.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c:435]: For interest -- section 5.2.4.2.1 of the C99 standard says that the `long` type must be at least 32 bits wide, so `strtoul()` (plus or minus some casts) seems like the way to go here.

[the remark this answers is not in the document]
Comment (sysvinit) [reply src/apps/relay/userdb.c]: Please see the revision I just pushed, this should hopefully have slightly cleaner integer parsing code.

File: src/apps/relay/userdb.c (hunk 7 above)
Comment (sysvinit) [reply src/apps/relay/userdb.c]: I've added static initialisation of the regex at startup in the revision I just pushed.

**Comment by supersven** [comment]
Hey @sysvinit ,

Sorry for the long delay. I'll have a look, now.

File: src/apps/relay/userdb.c [hunk 12]
```diff
@@ -395,6 +405,75 @@ static char *get_real_username(char *usname)
...
+	regmatch_t matches[ZREST_USERNAME_CAPTURES];
+	int ret;
+	char *span, *tmp, *endptr;
+	regoff_t span_len;
+
+	ret = -1;
+
+	/* XXX(sysvinit): this regex should not be repeatedly compiled, this
+	   could be very inefficient under load. */
+	if(regcomp(&reg, ZREST_USERNAME_REGEX, REG_EXTENDED)!=0) {
```
[the remark this answers is not in the document]
Comment (supersven) [reply src/apps/relay/userdb.c]: I was under the impression that the formatting style is mixed. However, we shouldn't waste more time on formatting. Finally, it's readable in both styles. :smiley:

File: src/apps/relay/userdb.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (supersven) [reply src/apps/relay/userdb.c]: Looks good :+1:

File: src/apps/relay/userdb.c (hunk 7 above)
Comment (supersven) [reply src/apps/relay/userdb.c]: Looks good :+1:

**Comment by supersven** [comment]
@sysvinit ,
I think this can be merged, now. :+1:

**Pull request merged**